include/aspect/simulator.h

```diff
@@ -459,6 +544,45 @@ namespace aspect
       void assemble_stokes_system ();
 
       /**
+       * Assemble and solve the temperature equation.
+       * This function returns the residual after solving
+       * and can optionally compute and store an initial
+       * residual before solving the equation.
+       *
+       * This function is implemented in
+       * <code>source/simulator/solver_schemes.cc</code>.
+       */
+      double assemble_and_solve_temperature (const bool compute_initial_residual = false,
+                                             double *initial_residual = NULL);
```

Comment:
The two arguments are duplicative: I think you can get rid of the first and document that the initial residual is computed if the pointer is non-`NULL`. Same for the other two functions of similar type.

Reply:
I do not quite agree. Of course I could replace 
`if (compute_initial_residual)` 
by 
`if(initial_residual!=NULL)`, 
but then I would need to call the function differently, namely instead of 
```
assemble_and_solve_temperature(nonlinear_iteration==0,initial_residual)
``` 
I would need to call 
```
assemble_and_solve_temperature((nonlinear_iteration==0) ?initial_residual : NULL))
```
, otherwise I would recompute the initial residual in every iteration. But then I would not have the initial residual available inside of the function in later iterations, so I could not return the relative residual, but only the absolute residual. This means I would need to move the conversion to relative residual into the individual solver schemes, which opens up the possibility for the output of solver schemes to diverge. I guess in the end both is possible, but I would prefer the current way. Do you insist on the change?

Reply:
No. I hadn't seen how you called the function. I think `(nonlinear_iteration==0) ?initial_residual : NULL)` would not be unreadable, but either is fine by me.

I did think about using `boost::optional` but decided that maybe that is going too far ;-)